### PR TITLE
Services ui

### LINKED ISSUE #1971: Triggering services from the UI
There is a backend call to trigger services

```
Suse::Backend.post("/source/@project/@package?cmd=runservice&user=@user")
```

This should be callable from `PackageController#show`. 

screenshot from 2016-07-28 13-44-57

### PR DESCRIPTION
**Pull request by Ana06**
Trigger services from the UI.

Closes #1971

### CODE REVIEW

File: src/api/app/controllers/webui/package_controller.rb [hunk 1]
```diff
@@ -663,6 +663,16 @@ def remove
     end
   end
 
+  def trigger_services
+    begin
+      Suse::Backend.post "/source/#{@project}/#{@package}?cmd=runservice&user=#{User.current}"
```
Comment (coolo) [inline src/api/app/controllers/webui/package_controller.rb]: you need to URI.escape those - devel:libraries:c++ will otherwise not be able to use this feature

File: src/api/spec/factories/packages.rb [hunk 2]
```diff
@@ -20,5 +20,12 @@
         end
       end
     end
+    factory :package_with_service do
+      after(:create) do |package|
+        # NOTE: Enable global write through when writing new VCR cassetes.
+        # ensure the backend knows the project
+        Suse::Backend.put("/source/#{CGI.escape(package.project.name)}/#{CGI.escape(package.name)}/_service", '<service/>')
```
Comment (coolo) [inline src/api/spec/factories/packages.rb]: This should actually be URI.escape - but I see yours is not the first spec to mix them ;(

File: src/api/spec/controllers/webui/package_controller_spec.rb [hunk 3]
```diff
@@ -518,4 +520,30 @@ def remove_file_post
...
+    before do
+      login user
+    end
+    
+    context 'with right params' do
+      before do
+        get :trigger_services, params: { project: source_project, package: service_package }
+      end
+      
+      it { expect(a_request(:post, "http://localhost:3200/source/#{source_project}/#{service_package}?cmd=runservice&user=#{user}")).to have_been_made.once }
```
Comment (ChrisBr) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: What about moving ``http://localhost:3200/source/`` to a let statement?

File: src/api/spec/controllers/webui/package_controller_spec.rb [hunk 4]
```diff
@@ -518,4 +520,30 @@ def remove_file_post
...
+      it { expect(flash[:notice]).to eq('Services successfully triggered') }
+      it { is_expected.to redirect_to(action: :show, project: source_project, package: service_package) }
+    end
+    
+    context "without a service file in the package" do
+      before do
+        get :trigger_services, params: { project: source_project, package: source_package }
+      end
+      
+      it { expect(a_request(:post, "http://localhost:3200/source/#{source_project}/#{source_package}?cmd=runservice&user=#{user}")).to have_been_made.once }
```
Comment (ChrisBr) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: What about moving ``http://localhost:3200/source/`` to a let statement?

File: src/api/app/controllers/webui/package_controller.rb [hunk 5]
```diff
@@ -1127,4 +1137,4 @@ def users_path
   def add_path(action)
     url_for(action: action, project: @project, role: params[:role], userid: params[:userid], package: @package)
   end
-end
+end
```
Comment (ChrisBr) [inline src/api/app/controllers/webui/package_controller.rb]: Huh?

File: src/api/spec/controllers/webui/package_controller_spec.rb [hunk 6]
```diff
@@ -1,4 +1,5 @@
 require 'rails_helper'
+require "webmock/rspec"
```
Comment (mdeniz) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: Why is this needed?

File: src/api/spec/controllers/webui/package_controller_spec.rb [hunk 7]
```diff
@@ -518,4 +520,32 @@ def remove_file_post
       end
     end
   end
+
+  describe 'GET #trigger_services' do
+    let(:post_url) { "http://localhost:3200/source/#{source_project}/#{service_package}?cmd=runservice&user=#{user}" }
```
Comment (mdeniz) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: I don't like using `localhost:3200` here, it can change an will break your tests

File: src/api/spec/controllers/webui/package_controller_spec.rb [hunk 8]
```diff
@@ -518,4 +520,32 @@ def remove_file_post
...
+        get :trigger_services, params: { project: source_project, package: service_package }
+      end
+
+      it { expect(a_request(:post, post_url)).to have_been_made.once }
+      it { expect(flash[:notice]).to eq('Services successfully triggered') }
+      it { is_expected.to redirect_to(action: :show, project: source_project, package: service_package) }
+    end
+
+    context "without a service file in the package" do
+      let(:post_url) { "http://localhost:3200/source/#{source_project}/#{source_package}?cmd=runservice&user=#{user}" }
```
Comment (mdeniz) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: Here the same...

File: src/api/app/controllers/webui/package_controller.rb [hunk 9]
```diff
@@ -663,6 +663,16 @@ def remove
     end
   end
 
+  def trigger_services
+    begin
+      Suse::Backend.post "/source/#{@project}/#{@package}?cmd=runservice&user=#{User.current}"
+      flash[:notice] = 'Services successfully triggered'
+    rescue RuntimeError, Timeout::Error, ActiveXML::Transport::NotFoundError, ActiveXML::Transport::Error
+      flash[:error] = "Services couldn't be triggered"
```
Comment (mdeniz) [inline src/api/app/controllers/webui/package_controller.rb]: No more information about the error?

File: src/api/app/views/webui/package/show.html.erb [hunk 10]
```diff
@@ -105,6 +105,9 @@
                 <li>
                   <%= link_to(sprited_text('package_delete', 'Delete package'), { :action => :delete_dialog, :package => @package, :project => @project }, :remote => true, id: 'delete-package') %>
                 </li>
+                <li>
+                  <%= link_to(sprited_text('package_link', 'Trigger services'), { action: :trigger_services, :package => @package, :project => @project }) %>
```
Comment (mdeniz) [inline src/api/app/views/webui/package/show.html.erb]: Why do you mix the old hash syntax with the new one?

File: src/api/config/routes.rb [hunk 11]
```diff
@@ -118,6 +118,7 @@ def self.matches?(request)
       post 'package/save_new_link/:project' => :save_new_link, constraints: cons
       post 'package/save/:project/:package' => :save, constraints: cons
       get 'package/delete_dialog/:project/:package' => :delete_dialog, constraints: cons
+      get 'package/trigger_services/:project/:package' => :trigger_services, constraints: cons
```
Comment (mdeniz) [inline src/api/config/routes.rb]: It should... in the end the call that is performed to the backend is a POST too.

File: src/api/app/controllers/webui/package_controller.rb (hunk 1 above)
Comment (mdeniz) [inline src/api/app/controllers/webui/package_controller.rb]: BTW, this is something I hope in the future we will avoid having a Suse::Backend library that "knows" how to run a service, instead of having to hand craft a url for it and just sending a POST request.

File: src/api/spec/controllers/webui/package_controller_spec.rb (hunk 6 above)
Comment (Ana06) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: It is needed for using `a_request`

File: src/api/spec/controllers/webui/package_controller_spec.rb (hunk 7 above)
Comment (Ana06) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: It is like that everywhere:

- https://github.com/openSUSE/open-build-service/blob/master/src/api/spec/features/webui/packages_spec.rb#L104

- https://github.com/openSUSE/open-build-service/blob/master/src/api/spec/models/package_spec.rb#L39

- https://github.com/openSUSE/open-build-service/blob/master/src/api/spec/models/service_spec.rb#L9

What is the alternative?

File: src/api/app/controllers/webui/package_controller.rb (hunk 5 above)
Comment (Ana06) [inline src/api/app/controllers/webui/package_controller.rb]: Rubocop complains :confused:

File: src/api/spec/controllers/webui/package_controller_spec.rb (hunk 7 above)
Comment (ChrisBr) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: @mdeniz is right, we should change the other places, too! 
Could you try if it works with a RegEx?

Comment (ChrisBr) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: And if it doesn't work I would prefer to only have one ``let`` statement with just the base url and concatenate it with path in the test.

File: src/api/spec/controllers/webui/package_controller_spec.rb (hunk 6 above)
Comment (mdeniz) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: Maybe it worth to move it into the rails_helper

File: src/api/spec/factories/packages.rb [hunk 12]
```diff
@@ -20,5 +20,12 @@
         end
       end
     end
```
Comment (bgeuken) [inline src/api/spec/factories/packages.rb]: Please add a whiteline here

File: src/api/spec/controllers/webui/package_controller_spec.rb (hunk 7 above)
Comment (Ana06) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: I don't know how I can use a regular expression for that, but I could use something like: 

`"http://#{CONFIG['source_host']}:#{CONFIG['source_port']}/source/#{source_project}/#{service_package}?cmd=runservice&user=#{user}"`

Is that fine?

File: src/api/spec/controllers/webui/package_controller_spec.rb (hunk 6 above)
Comment (Ana06) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: I am not sure of where do you want to move that and why... :confused:

File: src/api/app/controllers/webui/package_controller.rb [hunk 13]
```diff
@@ -663,6 +663,16 @@ def remove
     end
   end
 
+  def trigger_services
+    begin
+      Suse::Backend.post "/source/#{URI.escape(@project.name)}/#{URI.escape(@package.name)}?cmd=runservice&user=#{User.current}"
+      flash[:notice] = 'Services successfully triggered'
+    rescue RuntimeError, Timeout::Error, ActiveXML::Transport::NotFoundError, ActiveXML::Transport::Error => e
```
Comment (bgeuken) [inline src/api/app/controllers/webui/package_controller.rb]: @Ana06 RuntimeError would be generic exceptions, right? That would mean we might miss possible bugs.

File: src/api/spec/controllers/webui/package_controller_spec.rb [hunk 14]
```diff
@@ -518,4 +520,38 @@ def remove_file_post
...
+      let(:error) do
+        "Services couldn't be triggered: <status code=\"404\">\n  <summary>no source service defined!</summary>" \
+        "\n  <details>404 no source service defined!</details>\n</status>\n"
+      end
```
Comment (bgeuken) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: Please don't do that. It just makes the test harder to read.

File: src/api/spec/controllers/webui/package_controller_spec.rb (hunk 8 above)
Comment (bgeuken) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: True. This could be DRYed by moving it up to line 525

Comment (Ana06) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: why do you  want to move this to line 525? :confused:

File: src/api/spec/controllers/webui/package_controller_spec.rb (hunk 14 above)
Comment (Ana06) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: and what is the alternative? If I keep it in one line it is too long and Rubocop complains.

File: src/api/spec/controllers/webui/package_controller_spec.rb (hunk 6 above)
Comment (mdeniz) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: To use it widely... like shoulda_matchers

File: src/api/spec/controllers/webui/package_controller_spec.rb (hunk 14 above)
Comment (mdeniz) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: I guess that @bgeuken means using a let for having a literal that is used only one time in the specs...

File: src/api/spec/controllers/webui/package_controller_spec.rb (hunk 6 above)
Comment (ChrisBr) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: But then we will load for every test ....

File: src/api/spec/controllers/webui/package_controller_spec.rb (hunk 14 above)
Comment (bgeuken) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: Exactly. Plus it makes the test harder to read. And readability is quite important for tests.

File: src/api/app/controllers/webui/package_controller.rb (hunk 13 above)
Comment (bgeuken) [inline src/api/app/controllers/webui/package_controller.rb]: @Ana06 RuntimeError would be generic exceptions, right? That would mean we might miss possible bugs.

Comment (bgeuken) [inline src/api/app/controllers/webui/package_controller.rb]: @Ana06 Please fix. I accidentally clicked the approve link...

File: src/api/spec/controllers/webui/package_controller_spec.rb (hunk 8 above)
Comment (bgeuken) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: Because if you move it more often it's better to move it one layer up. Which in this case would be within "describe 'GET #trigger_services' do".

File: src/api/spec/controllers/webui/package_controller_spec.rb (hunk 14 above)
Comment (Ana06) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: `post_url` is also used only once. Do you also want me to remove it?
It will be difficult to read anyway, as it is a very long line.

Comment (bgeuken) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: I am not that opinionated about the post_url. But if we test the content of a flash message we would should do that right away. Not by juggling with some variables.

**Comment by Ana06** [comment]
@mdeniz @ChrisBr I have written the post url like this:

``` Ruby
"#{CONFIG['url']}/source/#{source_project}/#{source_package}?cmd=runservice&user=#{user}"
```
Is it fine like that? If so I'll change it in the rest of the places where `http://localhost:3200` is used. I'll do it in a different PR.

**Comment by mdeniz** [comment]
Ok to another PR with those changes.

But, CONFIG['url'] works?, BTW I prefer to use Configuration model

**Comment by coolo** [comment]
BTW: how do you handle the addition to release notes? This is quite a noteworthy feature

File: src/api/spec/controllers/webui/package_controller_spec.rb (hunk 8 above)
Comment (bgeuken) [inline src/api/spec/controllers/webui/package_controller_spec.rb]: Nevermind. Ana showed me that there is a small difference in the urls

File: src/api/config/routes.rb (hunk 11 above)
Comment (bgeuken) [inline src/api/config/routes.rb]: She changed it now

**Comment by bgeuken** [comment]
Nice work!

**Comment by bgeuken** [comment]
@Ana06 Please update the release notes as coolo mentioned

**Comment by mdeniz** [comment]
[(3) As a user I want to be able to trigger services from the UI to control when they are run](https://trello.com/c/45K5QWnu/32-3-as-a-user-i-want-to-be-able-to-trigger-services-from-the-ui-to-control-when-they-are-run)
